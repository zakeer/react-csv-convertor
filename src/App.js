import React, { useCallback, useState, useMemo } from 'react';
import {
  isCSVFile,
  getFileContent,
  convertCsvToJSON,
} from './utils';
import './style.css';

export default function App() {
  const [fileError, setFileError] = useState(false);
  const [csvData, setCsvData] = useState(null);

  const onFileChange = useCallback((event) => {
    const file = event.target.files[0];
    const isValidCSVFile = isCSVFile(file); // .html | !false -> true
    setFileError(!isValidCSVFile);
    console.log(file);

    if (isValidCSVFile) {
      getFileContent(file)
        .then((fileData) => {
          setCsvData(fileData);
        })
        .catch((fileError) =>
          console.log(':: ERROR READING FILE ::', fileError)
        );
    }

    // getFileContentByCallback(
    //   file,
    //   (fileData) => console.log(':: Callback based result ::', fileData),
    //   (fileError) => console.log(':: ERROR READING FILE ::', fileError)
    // );
  }, []);

  const { headers, data } = useMemo(() => {
    if (!csvData) {
      return {
        headers: [],
        data: [],
      };
    }
    return convertCsvToJSON(csvData);
    /*
    {
      headers,
      data,
    };
    */
  }, [csvData]);

  console.log(':: HEADER AND DATA ::', { headers, data });

  return (
    <div>
      <h1>CSV TO Table</h1>
      <p>
        Please upload <strong>.csv</strong> to see some magic happen :)
      </p>

      {fileError && (
        <p style={{ color: 'red' }}>Please provide correct .csv file </p>
      )}

      <input data-testid="inputFile" type="file" accept=".csv" onChange={onFileChange} />

      <table border="1">
        <thead>
          <tr>
            {headers.map((header) => (
              <th key={header}>{header}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {data.map((columns, idx) => (
            <tr key={idx}>
              {columns.map((column) => (
                <td key={column}>{column}</td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}
