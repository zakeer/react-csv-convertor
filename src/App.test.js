import { fireEvent, render, screen, act, waitFor } from '@testing-library/react';
import React from 'react';
import App from './App';


describe("App Component", () => {
  test("Should render successfully", () => {
    const appComponent = render(<App />);
    expect(appComponent).toMatchSnapshot();
  })

  test("Process CSV to Table", async () => {

    const FILE_CONTENT = `Name,Job
    Lakshmi,  UI Engineer
    Kalyan, DevOPS
    Haseena, React Developer`;

    const FILE_NAME = 'data.csv';

    const readAsTextMock = jest.fn();
    jest.spyOn(global, 'FileReader').mockImplementation(function () {
      const self = this;
      this.readAsText = readAsTextMock.mockImplementation(() => {
        self.onload({ target: { result: FILE_CONTENT } });
      });
    });

    const appComponent = render(<App />);
    const input = screen.getByTestId("inputFile"); // <element data-testid="inputFile"  />
    expect(input).toBeInTheDocument();

    const file = new File([FILE_CONTENT], FILE_NAME);

    act(() => {
      fireEvent.change(input, {
        target: {
          files: [file]
        }
      })
    });
    
    await waitFor(() => expect( screen.getByText('UI Engineer') ).toBeInTheDocument())
    expect(appComponent).toMatchSnapshot();

  })


})
