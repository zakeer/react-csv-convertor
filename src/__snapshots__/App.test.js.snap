// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Process CSV to Table 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          CSV TO Table
        </h1>
        <p>
          Please upload 
          <strong>
            .csv
          </strong>
           to see some magic happen :)
        </p>
        <input
          accept=".csv"
          data-testid="inputFile"
          type="file"
        />
        <table
          border="1"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Job
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                Lakshmi
              </td>
              <td>
                UI Engineer
              </td>
            </tr>
            <tr>
              <td>
                Kalyan
              </td>
              <td>
                DevOPS
              </td>
            </tr>
            <tr>
              <td>
                Haseena
              </td>
              <td>
                React Developer
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        CSV TO Table
      </h1>
      <p>
        Please upload 
        <strong>
          .csv
        </strong>
         to see some magic happen :)
      </p>
      <input
        accept=".csv"
        data-testid="inputFile"
        type="file"
      />
      <table
        border="1"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Job
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Lakshmi
            </td>
            <td>
              UI Engineer
            </td>
          </tr>
          <tr>
            <td>
              Kalyan
            </td>
            <td>
              DevOPS
            </td>
          </tr>
          <tr>
            <td>
              Haseena
            </td>
            <td>
              React Developer
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App Component Should render successfully 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          CSV TO Table
        </h1>
        <p>
          Please upload 
          <strong>
            .csv
          </strong>
           to see some magic happen :)
        </p>
        <input
          accept=".csv"
          data-testid="inputFile"
          type="file"
        />
        <table
          border="1"
        >
          <thead>
            <tr />
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        CSV TO Table
      </h1>
      <p>
        Please upload 
        <strong>
          .csv
        </strong>
         to see some magic happen :)
      </p>
      <input
        accept=".csv"
        data-testid="inputFile"
        type="file"
      />
      <table
        border="1"
      >
        <thead>
          <tr />
        </thead>
        <tbody />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
